va/wayland/va_wayland_linux_dmabuf.c

```diff
@@ -0,0 +1,364 @@
+        if (dev != match)
+            drmFreeDevice(&dev);
+    }
+    free(devices);
+
+    return match;
+}
+
+static void
+feedback_handle_main_device(
+    void                                *data,
+    struct zwp_linux_dmabuf_feedback_v1 *feedback,
+    struct wl_array                     *device_array
+)
+{
+    dev_t dev_id;
+    drmDevice *dev;
+    const char *dev_path;
+    VADisplayContextP const pDisplayContext = data;
+    VADriverContextP const ctx = pDisplayContext->pDriverContext;
+    struct drm_state * const drm_state = ctx->drm_state;
+
+    assert(device_array->size == sizeof(dev_id));
+    memcpy(&dev_id, device_array->data, sizeof(dev_id));
+
+    dev = get_drm_device_from_dev_id(dev_id);
+    if (!dev) {
+        va_wayland_error("failed to get DRM device from device ID");
+        return;
+    }
+
+    if (!(dev->available_nodes & (1 << DRM_NODE_RENDER)))
+        goto end;
+
+    dev_path = dev->nodes[DRM_NODE_RENDER];
+    drm_state->fd = open(dev_path, O_RDWR | O_CLOEXEC);
+    if (drm_state->fd < 0) {
+        va_wayland_error("failed to open %s", dev_path);
+        goto end;
+    }
+
+    drm_state->auth_type = VA_DRM_AUTH_CUSTOM;
+
+end:
+    drmFreeDevice(&dev);
+}
+
+static void
+feedback_handle_tranche_done(
+    void                                *data,
+    struct zwp_linux_dmabuf_feedback_v1 *feedback
+)
+{
+}
+
+static void
+feedback_handle_tranche_target_device(
+    void                                *data,
+    struct zwp_linux_dmabuf_feedback_v1 *feedback,
+    struct wl_array                     *device_array
+)
+{
+}
+
+static void
+feedback_handle_tranche_formats(
+    void                                *data,
+    struct zwp_linux_dmabuf_feedback_v1 *feedback,
+    struct wl_array                     *indices_array
+)
+{
+}
+
+static void
+feedback_handle_tranche_flags(
+    void                                *data,
+    struct zwp_linux_dmabuf_feedback_v1 *feedback,
+    uint32_t                             flags
+)
+{
+}
+
+static const struct zwp_linux_dmabuf_feedback_v1_listener feedback_listener = {
+    .done = feedback_handle_done,
+    .format_table = feedback_handle_format_table,
+    .main_device = feedback_handle_main_device,
+    .tranche_done = feedback_handle_tranche_done,
+    .tranche_target_device = feedback_handle_tranche_target_device,
+    .tranche_formats = feedback_handle_tranche_formats,
+    .tranche_flags = feedback_handle_tranche_flags,
+};
+
+static void
+registry_handle_global(
+    void               *data,
+    struct wl_registry *registry,
+    uint32_t            name,
+    const char         *interface,
+    uint32_t            version
+)
+{
+    VADisplayContextP const pDisplayContext = data;
+    struct va_wayland_linux_dmabuf_context *wl_linux_dmabuf_ctx = pDisplayContext->opaque;
+    struct zwp_linux_dmabuf_v1 *linux_dmabuf;
+    struct zwp_linux_dmabuf_feedback_v1 *feedback;
+
+    if (strcmp(interface, zwp_linux_dmabuf_v1_interface.name) == 0 &&
```

Comment:
TBH, I am not familiar with this interface, but from code logic, 
looks , it could be combined together with va_wayland_drm? add interface check inside registry_handle_global
if it is linuxdmabuf ,xxxx ; if it is wl_drm xxxx?

from libva perspective: it should retrieve the device fd though these function, then get the driver name though device fd.  looks both wl_drm and linuxdma buffer could do it.  but wl_drm interface will be deprecated?

from driver implementation perspective. what's the difference to implement vaGetSurfaceBufferWl and vaGetImageBufferWl? previously, we use "wl_proxy_create" and "wl_proxy_marshal" interface to share the dma buffer fd from vaSurface to wl_buffer, just like https://github.com/intel/intel-vaapi-driver/blob/master/src/i965_output_wayland.c#L231-L263.  looks we need new implementation now?

Reply:
> looks , it could be combined together with va_wayland_drm? add interface check inside registry_handle_global
if it is linuxdmabuf ,xxxx ; if it is wl_drm xxxx?

That's possible indeed, but not sure it's a good idea. It would entangle the two implementations and make it more complicated to understand the code and drop the `wl_drm` part in the future.

> from libva perspective: it should retrieve the device fd though these function, then get the driver name though device fd. looks both wl_drm and linuxdma buffer could do it. but wl_drm interface will be deprecated?

Yes, `wl_drm` was never supposed to be used by clients such as libva, it was supposed to be an internal Mesa protocol.

Some Wayland compositors like Sway have already dropped support for `wl_drm`.

> from driver implementation perspective. what's the difference to implement vaGetSurfaceBufferWl and vaGetImageBufferWl? previously, we use "wl_proxy_create" and "wl_proxy_marshal" interface to share the dma buffer fd from vaSurface to wl_buffer, just like https://github.com/intel/intel-vaapi-driver/blob/master/src/i965_output_wayland.c#L231-L263. looks we need new implementation now?

Hm, yeah. This would need a new implementation indeed.

mpv seems to work just fine without this new implementation though. Maybe because they grab the DMA-BUFs directly, instead of `vaGetImageBufferWl`/`vaGetSurfaceBufferWl`?

There is nothing driver-specific in `vaGetSurfaceBufferWl`/`vaGetImageBufferWl` by the way. The implementation should be identical for all VA-API drivers.

Reply:
FWIW, it seems like Mesa doesn't implement `VADriverVTableWayland`.